File: Programs/matlab/mPLDA/mPLDA_GroupScoring5.m
%=========================================================
% function [scores, clusterID] = mPLDA_GroupScoring1(mPLDAModel, Xs, xt)
% Implement the SNR-independent mixture of PLDA scoring. This function should
% work with mPLDA_train5.m
%
% This function is for opt.mode = 'scravg' and 'ivcsnravg' in snr3_score_gplda_w.m.
% When mode = 'icvsnravg', the SNR within the same group must be averaged and the
% length of Ls must be equal to the number of SNR groups.
% 
%   Input:
%       mPLDAModel     - mPLDA model structure
%       Xs             - Matrix containing a set of column i-vectors of speaker s
%       xt             - Second un-normalized i-vector (column vec)
%   Output:
%       scores         - PLDA scores (unnormalized) of Xs and xt
% Author: M.W. Mak
% Date: Aug 2015
%=========================================================
function scores = mPLDA_GroupScoring5(mPLDAModel, Xs, xt)

% Whitening/WCCN then length normalization
n_vecs = size(Xs,2);
Xs = mPLDAModel.projmat1' * (Xs-repmat(mPLDAModel.meanVec1,1,n_vecs));
xt = mPLDAModel.projmat1' * (xt-mPLDAModel.meanVec1);
Xs = (len_norm(Xs'))';
xt = (len_norm(xt'))';
Xs = mPLDAModel.projmat2' * Xs;
xt = mPLDAModel.projmat2' * xt;

% Extract paras from model structure
varphi = mPLDAModel.varphi;
m = mPLDAModel.m;
Icov = mPLDAModel.Icov;
Icov2 = mPLDAModel.Icov2;
logDetCov = mPLDAModel.logDetCov;
logDetCov2 = mPLDAModel.logDetCov2;
V = mPLDAModel.V;
K = length(varphi);

% Precompute likelihood of test i-vector for speed
sum3 = 0;
expterm = zeros(K,1);
for k = 1:K,
    detterm = 0.5*logDetCov{k};
    expterm(k) = exp(-0.5*Mahaldist(xt, m{k}, Icov{k})-detterm); 
    sum3 = sum3 + varphi(k)*expterm(k);
end
assert(sum3>0,'Divided by zero (sum3) in mPLDA_GroupScoring.m');

% Compute log-likelihood score
scores = zeros(size(Xs,2),1);
for s = 1:length(scores),
    sum1 = 0; sum2 = 0; 
    xs = Xs(:,s);
    for ks = 1:K,
        for kt = 1:K,
            sum1 = sum1 + varphi(ks)*varphi(kt) * ...
                   exp(-0.5*Mahaldist([xs; xt],[m{ks}; m{kt}],Icov2{ks,kt})-0.5*logDetCov2{ks,kt});
        end
    end
    for k = 1:K,
        detterm = 0.5*logDetCov{k};
        sum2 = sum2 + varphi(k)*exp(-0.5*Mahaldist(xs, m{k}, Icov{k})-detterm); 
    end
    assert(sum1>0,'Sum1 is 0 in mPLDA_GroupScoring.m');
    assert(sum2>0,'Divided by zero (sum2) in mPLDA_GroupScoring.m');
    scores(s) = log(sum1) - (log(sum2)+log(sum3));
end
    
    

%% Private functions

% Return the Mahalanobis distance between x and mu with covariance Sigma
% Both x and mu are col vectors
function md = Mahaldist(x, mu, Icov)
temp = x - mu;
md = (temp'*Icov)*temp;




        

File: Programs/python/mPLDA/mPLDA_GroupScoring5_.py
import numpy as np
# import cupy as np
from len_norm_ import len_norm

def Mahaldist(x, mu, Icov):
    """
     Return the Mahalanobis distance between x and mu with covariance Sigma
     Both x and mu are col vectors
    """

    temp = x - mu
    md = np.dot(np.dot(temp.T, Icov), temp)
    return md

def mPLDA_GroupScoring5(mPLDAModel, Xs, xt):
    """
     function [scores, clusterID] = mPLDA_GroupScoring1(mPLDAModel, Xs, xt)
     Implement the SNR-independent mixture of PLDA scoring. This function should
     work with mPLDA_train5_.py

     This function is for opt.mode = 'scravg' and 'ivcsnravg' in snr3_score_gplda_w_.py.
     When mode = 'icvsnravg', the SNR within the same group must be averaged and the
     length of Ls must be equal to the number of SNR groups.

       Input:
           mPLDAModel     - mPLDA model structure
           Xs             - Matrix containing a set of column i-vectors of speaker s
           xt             - Second un-normalized i-vector (column vec)
       Output:
           scores         - PLDA scores (unnormalized) of Xs and xt
     Author: M.W. Mak
     Date: Aug 2015
    """

    n_vecs = Xs.shape[1]
    Xs = np.dot(mPLDAModel.projmat1.T, Xs - np.tile(mPLDAModel.meanVec1, [1, n_vecs]))
    xt = np.dot(mPLDAModel.projmat1.T, xt - mPLDAModel.meanVec1)
    Xs = len_norm(Xs.T).T
    xt = (len_norm(xt.T)).T
    Xs = np.dot(mPLDAModel.projmat2.T, Xs)
    xt = np.dot(mPLDAModel.projmat2.T, xt)

    # Extract paras from model structure
    varphi = mPLDAModel.varphi
    m = mPLDAModel.m
    Icov = mPLDAModel.Icov
    Icov2 = mPLDAModel.Icov2
    logDetCov = mPLDAModel.logDetCov
    logDetCov2 = mPLDAModel.logDetCov2
    V = mPLDAModel.V
    K = len(varphi)

    # Precompute likelihood of test i-vector for speed
    sum3 = 0
    expterm = np.zeros(shape=(K, 1), dtype=np.float64)
    for k in range(K):
        detterm = 0.5 * logDetCov[0][k]
        expterm[k, 0] = np.exp(-0.5 * Mahaldist(xt, m[0][k], Icov[0][k]) - detterm)
        sum3 = sum3 + varphi[k, 0] * expterm[k, 0]

    assert sum3 > 0, 'Divided by zero (sum3) in mPLDA_GroupScoring_.py'

    # Compute log-likelihood score
    scores = np.zeros(shape=(Xs.shape[1], 1), dtype=np.float64)
    for s in range(len(scores)):
        sum1 = 0
        sum2 = 0
        xs = Xs[:, s].reshape(-1, 1)
        for ks in range(K):
            for kt in range(K):
                sum1 = sum1 + varphi[ks][0] * varphi[kt][0] * \
                   np.exp(-0.5 * Mahaldist(np.vstack((xs, xt)), np.vstack((m[0][ks], m[0][kt])), Icov2[ks][kt]) - 0.5 * logDetCov2[ks][kt])
        for k in range(K):
            detterm = 0.5 * logDetCov[0][k]
            sum2 = sum2 + varphi[k][0] * np.exp(-0.5 * Mahaldist(xs, m[0][k], Icov[0][k]) - detterm)
        assert sum1 > 0, 'Sum1 is 0 in mPLDA_GroupScoring_.py'
        assert sum2 > 0, 'Divided by zero (sum2) in mPLDA_GroupScoring_.py'
        scores[s][0] = np.log(sum1) - (np.log(sum2) + np.log(sum3))
    return scores